quickwit/quickwit-indexing/src/source/pulsar_source.rs

```diff
@@ -0,0 +1,1083 @@
+                    self.process_message(message, &mut batch).map_err(ActorExitStatus::from)?;
+
+                    if batch.num_bytes >= BATCH_NUM_BYTES_LIMIT {
+                        break;
+                    }
+                }
+                _ = &mut deadline => {
+                    break;
+                }
+            }
+            ctx.record_progress();
+        }
+
+        if !batch.checkpoint_delta.is_empty() {
+            debug!(
+                num_docs=%batch.docs.len(),
+                num_bytes=%batch.num_bytes,
+                num_millis=%now.elapsed().as_millis(),
+                "Sending doc batch to indexer.");
+            let message = batch.build();
+            ctx.send_message(doc_processor_mailbox, message).await?;
+        }
+
+        Ok(Duration::default())
+    }
+
+    async fn suggest_truncate(
+        &self,
+        checkpoint: SourceCheckpoint,
+        _ctx: &ActorContext<SourceActor>,
+    ) -> anyhow::Result<()> {
+        debug!(ckpt = ?checkpoint, "Truncating message queue.");
+        let _ = self.pulsar.last_ack.send(checkpoint);
+        Ok(())
+    }
+
+    fn name(&self) -> String {
+        format!(
+            "PulsarSource{{source_id={}}}",
+            self.ctx.source_config.source_id
+        )
+    }
+
+    fn observable_state(&self) -> JsonValue {
+        json!({
+            "index_id": self.ctx.index_id,
+            "source_id": self.ctx.source_config.source_id,
+            "topics": self.params.topics,
+            "subscription_name": self.subscription_name,
+            "consumer_name": self.params.consumer_name,
+            "num_bytes_processed": self.state.num_bytes_processed,
+            "num_messages_processed": self.state.num_messages_processed,
+            "num_invalid_messages": self.state.num_invalid_messages,
+        })
+    }
+}
+
+#[derive(Debug)]
+struct PulsarMessage;
+
+impl DeserializeMessage for PulsarMessage {
+    type Output = anyhow::Result<String>;
+
+    fn deserialize_message(payload: &Payload) -> Self::Output {
+        String::from_utf8(payload.data.clone()).map_err(anyhow::Error::from)
+    }
+}
+
+#[derive(Debug, Default)]
+struct BatchBuilder {
+    docs: Vec<String>,
+    num_bytes: u64,
+    checkpoint_delta: SourceCheckpointDelta,
+}
+
+impl BatchBuilder {
+    fn build(self) -> RawDocBatch {
+        RawDocBatch {
+            docs: self.docs,
+            checkpoint_delta: self.checkpoint_delta,
+        }
+    }
+
+    fn push(&mut self, doc: String, num_bytes: u64) {
+        self.docs.push(doc);
+        self.num_bytes += num_bytes;
+    }
+}
+
+struct PulsarConsumer {
+    messages: mpsc::Receiver<ConsumerMessage>,
+    last_ack: watch::Sender<SourceCheckpoint>,
+}
+
+#[tracing::instrument(name = "pulsar-consumer", skip(pulsar))]
+/// Spawns a background task listening for incoming messages to
+/// process.
+///
+/// The exists because the `Consumer` is not `Sync` or `Send` in places which means we cannot
```

Comment:
I should re-word this comment/doc, the `Consumer` type itself is `Send` but seeking and calling the seeking method is *not* `Send` nor `Sync` which when investigating looked to be due to the fact that `pulsar-rs` makes use of some dynamic dispatch and async streams which do not seem to play nicely with the seek method.

I did attempt to use a Mutex (specifically a tokio::sync::Mutex) but this did not satisfy the compiler, other attempts also include writing a separate actor for the consumer using a tokio task, but this limitation requires a local set to guarantee the future stays on the same thread.

For replicating the issue you specifically need to call `seek` as that is what creates the `!Send` future from what I can tell, which under the hood is re-constructing the clients and replacing itself after seeking which I imagine is where the issue lies.

Reply:
Instead of running the whole consumer on a dedicated thread, let's run the `!Send` future in a dedicated runtime pinned to a thread and use a local set. At least we can share that runtime between multiple Pulsar source instances. Otherwise, we require one thread per source instance, which is not great (and an issue for the Kafka source).

Reply:
> Instead of running the whole consumer on a dedicated thread, let's run the !Send future in a dedicated runtime pinned to a thread and use a local set. At least we can share that runtime between multiple Pulsar source instances. Otherwise, we require one thread per source instance, which is not great (and an issue for the Kafka source).

I also tried this but was unsuccessful in getting this approach to work. 

The mid-ground would be we use tokio-utils'  local set pool which will let use share this runtime.

Reply:
For future reference, [here](https://gist.github.com/guilload/55fab910d11b7eb23aef4d3af227279f) is how to do it.

... or https://github.com/streamnative/pulsar-rs/pull/255

Reply:
> https://github.com/streamnative/pulsar-rs/pull/255

Looks like that's the way to go then if we can get that PR merged.